rts/Sim/Units/UnitTypes/Builder.cpp

```diff
@@ -729,6 +730,18 @@ bool CBuilder::StartBuild(BuildInfo& buildInfo, CFeature*& feature, bool& inWait
 
 	buildInfo.pos = CGameHelper::Pos2BuildPos(buildInfo, true);
 
+	auto isBuildeeFloating = [](BuildInfo& buildInfo) {
+		if (buildInfo.def->RequireMoveDef()) {
+			MoveDef* md = moveDefHandler.GetMoveDefByPathType(buildInfo.def->pathType);
+			return (md->FloatOnWater());
+		} else {
+			return (buildInfo.def->floatOnWater);
+		}
+	};
+
+	if (isBuildeeFloating(buildInfo))
+		buildInfo.pos.y = (buildInfo.pos.y < 0.f) ? 0.f : buildInfo.pos.y;
```

Comment:
why is this only a problem for floaters? why doesn't, for example, setting `y` to 9999 also fail (unit at ground level won't overlap with one nominally in the sky)? a comment about there being floater-specific logic would be good, if we can't remove the floater-specific logic itself.

Reply:
Yes. A comment would help here. It isn't obvious that the engine has special collision checks for units in water.